### PR TITLE
Read metadata when creating from existing feed

### PR DESCRIPTION
**Pull request by timgoeller**
Currently the metadata field is not filled when the hyperbee is created from an existing non-writable feed. This commit retrieves the header from the feed, and sets the metadata field. This would cause the hyperbee to not resolve until the first feed entry is loaded, but one could argue that the tree is not really ready when there is no header. It might make sense to also check if the protocol is set to `hyperbee`.

### CODE REVIEW

**Comment by timgoeller** [comment]
I've experimented with decoding the header in my project, and it might make sense to not block `ready` so there's no deadlock because the module using the hyperbee waits for `ready` to start replication, but instead trigger a different event once the header was read successfully (maybe something like `initialized` or `valid`).

**Comment by mafintosh** [comment]
Thanks for the PR, will and review this week :)

**Comment by mafintosh** [comment]
Ya, you shouldn't block the ready handler. If you use get wait: false it'll fail immediately if it doesn't have the data.

**Comment by timgoeller** [comment]
Huh, yeah, that would be an option. Though I think an event that triggers once/if the get is completed would be nicer, because then you could wait for that to happen if you wanted to. Otherwise it would fail and you'd still need to read the header manually at some point. What do you think? @mafintosh

**Comment by mafintosh** [comment]
@timgoeller ya, you can that doing something similar to this:

``` js
ready (cb) {
  feed.get(0, { wait: false }, (_, data) => {
    if (data) setMetadata(data) // we had it locally, set metadata
    else feed.get(0, { wait: true }, (_, data) => setMetadata(data)) // we got it later, set it 
    cb()
  })
}
```

**Comment by timgoeller** [comment]
I removed the code from `_open()`, so it won't block `ready` in any way. Instead I created a method to retrieve the metadata if needed. If it's already set because the hyperbee was just created (or the data was already retrieved) it will just resolve with the value. Otherwise it will try to get it, and resolve once `get` finishes. I think in that case it probably (?) would make no difference to call it `wait: false` and then with `wait: true` again.

This way I can get the metadata if I want to, but it won't block anything else. 

One thing I noticed though: When the feed is created without metadata, `Header.decode()` will throw an error (Error: Decoded message is not valid). Not sure if that's a bug.

File: index.js [hunk 1]
```diff
@@ -302,6 +302,19 @@ class HyperBee {
     })
   }
 
+  retrieveMetadata () {
+    return new Promise((resolve, reject) => {
+      if (this.metadata !== null) resolve(this.metadata)
+
+      this.feed.get(0, (err, data) => {
+        if (err) reject(err)
```
Comment (mafintosh) [inline index.js:310]: return reject(err)

File: index.js [hunk 2]
```diff
@@ -302,6 +302,19 @@ class HyperBee {
     })
   }
 
+  retrieveMetadata () {
```
Comment (mafintosh) [inline index.js:305]: can we call this getMetadata ?

File: index.js [hunk 3]
```diff
@@ -302,6 +302,19 @@ class HyperBee {
...
+    return new Promise((resolve, reject) => {
+      if (this.metadata !== null) resolve(this.metadata)
```
Comment (mafintosh) [inline index.js:307]: return resolve(this.metadata)

File: index.js (hunk 2 above)
Comment (timgoeller) [reply index.js:305]: sure, my only concern there was that it might cause confusion because of the `metadata` field

**Comment by mafintosh** [comment]
Hi, we have `.getHeader()` now :)

**Pull request closed without merging**